File: leetcode_balanced_binary_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Boolean}

def is_balanced(root)
    return check_recursive(root)
end

def check_recursive(node)
    return 0 unless node

    height_right = 1 + check_recursive(node.right)
    height_left = 1 + check_recursive(node.left)

    return -1 if height_left == -1 || height_right == -1
    return -1 if if (height_left - height_right).abs > 1
    return [height_left, height_right].max
end

File: LEETCODE_balanced_binary_tree.cs
public class Solution {
    public bool IsBalanced(TreeNode root) {
        if (root == null) return true;

        var left = Height(root.left);
        var right = Height(root.right);

        if (Math.Abs(left - right) > 1) return false;

        return IsBalanced(root.left) && IsBalanced(root.right);
    }

    private int Height(TreeNode node)
    {
        if (node == null) return 0;

        int left = Height(node.left);
        int right = Height(node.right);

        return 1 + Math.Max(left, right);
    }
}
